File: code/ex68.java
class Solution {
    public List<String> fullJustify(String[] words, int maxWidth) {
        int n = words.length;
        List<String> res = new ArrayList<>();
        
        for (int i = 0, l, k; i < n; i += k) {
            // Get the max number of words can be put in current line
            for (l = k = 0; i + k < n && l + k + words[i + k].length() <= maxWidth; k++) {
                l += words[i + k].length();
            }
            
            // Build current line
            String line = words[i];
            for (int j = 1; j < k; j++) {
                if (i + k >= n) {    // The last line
                    line += " ";                    
                } else {
                    int spaces = (maxWidth - l) / (k - 1) + ((j <= (maxWidth - l) % (k - 1)) ? 1 : 0);
                    line += multipleSpaces(spaces);
                }
                line += words[i + j];
            }
            int restSpaces = maxWidth - line.length();
            line += multipleSpaces(restSpaces);
            res.add(line);
        }
        
        return res;
    }
    
    private String multipleSpaces(int n) {
        String ret = "";
        for (int i = 0; i < n; i++)
            ret += " ";
        return ret;
    }
}

File: code/ex68.cpp
class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
        int n = words.size();
        vector<string> res;

        for (int i = 0, k, l; i < n; i += k) {    // k => cnt of words on current line
            // Get the cnt of words on current line
            for (k = l = 0; i + k < n && k + l + words[i + k].size() <= maxWidth; k++) {
                l += words[i + k].length();
            }
            
            // Build current line
            string cur = words[i];
            for (int j = 1; j < k; j++) {
                // Adding spaces between words
                if (i + k >= n) {    // Last line => left-justified
                    cur += " ";
                } else {
                    int spaces = (maxWidth - l) / (k - 1) + (j <= (maxWidth - l) % (k - 1) ? 1 : 0);
                    cur += string(spaces, ' ');
                }
                cur += words[i + j];
            }
            cur += string(maxWidth - cur.length(), ' ');
            res.push_back(cur);
        }

        return res;
    }
};